FEAT: [#34] 장바구니 담기 API 구현

no-cy opened the pull request:
### 관련 이슈
- #34 

### 내용
- 장바구니 담기 API 구현
- 장바구니 담기 API 테스트 코드 작성
- 엔티티 객체의 데이터 일관성을 위해 `@Setter` 삭제 후 build 추가

close #34

no-cy committed:
[#34] 장바구니 담기 API 리포지토리 구현

- CartRepository
  - findCartByUserID: 유저의 장바구니 정보 조회 메서드
  - countProductInCart: 장바구니에 담으려는 상품이 장바구니에 존재하는지 확인하기 위한 메서드
- CartItemRepository
  - updateProductQuantity: 상품이 이미 장바구니에 존재하면, 상품 수량을 증가시키는 메서드

no-cy committed:
[#34] 장바구니 담기 API 리포지토리 테스트 코드 작성

no-cy committed:
[#34] 장바구니 담기 API 서비스 구현 및 테스트 코드 작성

- 조회하고자하는 유저의 장바구니와 상품이 없을 경우 EntityNotFoundException이 발생되도록 구현
- 상품을 장바구니에 추가 시, 상품이 이미 장바구니에 있을 경우 상품 수량이 증가될 수 있도록 로직 구현
- 장바구니 담기 서비스 테스트 코드 작성

no-cy committed:
[#34] 장바구니 담기 API 컨트롤러 구현 및 Dto 생성

- AddProductToCartRequest, AddProductToCartResponse 구현
- 요청 받은 수량이 0개 이하일 경우 500번 에러를 전달하도록 구현
- 유저의 장바구니와 상품을 조회하지 못할 경우에도 500번 에러를 전달하도록 구현

no-cy committed:
[#34] 장바구니 담기 API 빌더 및 테스트 코드구현

- 데이터의 일관성을 위해 엔티티 객체에 @Setter를 삭제하고, builder 추가
- builder 테스트 코드 작성

no-cy committed:
[#34] 장바구니 담기 API 테스트 코드 public 접근 제어자 삭제 (Junit5)

- Junit5에서는 리플렉션을 사용하여 내부적으로 테스트 메서드에 접근하기 때문에 'public'으로 지정할 필요없음
- 이로써, 패키지 내부에서만 접근 가능하도록 설정됨
- 필요 이상의 접근 범위를 제한하여 캡슐화 강화

no-cy committed:
[#34] 장바구니 담기 API 테스트 코드 public 접근 제어자 추가 삭제

f-lab-kane commented on src/main/java/spring/marketnori/cart/Cart.java:
```diff
@@ -1,14 +1,16 @@
 @Getter
-@Setter
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
 public class Cart extends Auditable {
```
찬영님 저번 멘토링때 언급드렸던 내용인데 저희 프로젝트 entity에서도 학습하면서 적용해봐요
https://github.com/f-lab-edu/market-nori/issues/40 이슈로 간단히 생성해두었습니다. 
이번 pr 말고 다음 pr에서 적용해봐요

no-cy replied:
넵 알겠습니다 ! 적용해 보도록 하겠습니다 !

f-lab-kane commented on src/main/java/spring/marketnori/cart/CartController.java:
```diff
@@ -0,0 +1,46 @@
+        AddProductToCartResponse addProductToCartResponse = new AddProductToCartResponse();
+        Long quantity = request.getQuantity();
+
+        if (quantity < 0) {
+            log.debug("The product quantity has not been specified correctly.");
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
+        }
```
spring bean validation 사용하면 코드가 한층더 간결하고 깔끔해 집니다.
참고 링크 드립니다. https://velog.io/@woply/spring-%EC%8A%A4%ED%94%84%EB%A7%81%EC%9D%B4-%EC%A7%80%EC%9B%90%ED%95%98%EB%8A%94-Bean-Validator-%EA%B8%B0%EB%8A%A5-%EC%82%AC%EC%9A%A9-%EB%B0%A9%EB%B2%95

no-cy replied:
[9acd61b](https://github.com/f-lab-edu/market-nori/pull/39/commits/9acd61b69b1b842d1121273b4aaf78e0d174317c)
멘토님, 공유해 주신 내용 참고하여 수정 완료하였습니다. 감사합니다.

f-lab-kane commented on src/main/java/spring/marketnori/cart/CartRepository.java:
```diff
@@ -1,6 +1,15 @@
 
 public interface CartRepository extends JpaRepository<Cart, Long> {
+    @Query("SELECT c FROM Cart c WHERE c.user.userId = :userId")
+    Optional<Cart> findCartByUserId(Long userId);
```
query method를 사용했어도 되었을거 같은데 jpql 쿼리를 별도로 작성해서 사용하신 이유가 있을까요?

no-cy replied:
멘토님, 멘토링 시간에 말씀해 주셨던 대로 Cart 엔티티에 userId 필드가 존재하지 않아 연관관계 엔티티인 User를 참조하여 userId를 가져와야 하므로 쿼리 메서드를 사용하지 못하였습니다.

[141448b](https://github.com/f-lab-edu/market-nori/pull/39/commits/141448bc27be2a25ac0b66b3231391a6fc2ab99d)
Cart에 userId 필드를 읽기 전용으로 생성함으로써 JPQL이 아닌 단순 쿼리 메서드로 Cart 객체를 조회할 수 있게 되었습니다.

f-lab-kane commented on src/main/java/spring/marketnori/cartitem/CartItemRepository.java:
```diff
@@ -1,6 +1,11 @@
+import org.springframework.data.jpa.repository.Query;
 
 public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Modifying
+    @Query("UPDATE CartItem ci SET ci.quantity = ci.quantity + :quantity WHERE ci.cart.cartId = :cartId AND ci.product.productId = :productsId AND ci.quantity > 0")
+    void updateProductQuantity(Long cartId, Long productsId, Long quantity);
```
jpql로 update할땐 AuditingEntityListener가 동작하지 않을수 있습니다. 
지금은 아직 적용하지 않으신듯 한데 나중에 적용하실때 확인해보시길 바래요~

no-cy replied:
넵 추후 Auditing 구현 후 해당 사항 확인해 보도록 하겠습니다 ! 감사합니다 !

f-lab-kane reviewed (approved):
수고하셨습니다 👍

no-cy committed:
[#34] 장바구니 담기 API @validated, @valid 어노테이션 추가

- 장바구니에 담을 상품 갯수에 대한 검증을 if 조건문이 아닌 @validated 어노테이션을 통해 검증을 수행하도록 변경

no-cy committed:
[#34] Cart 엔티티에 읽기 전용 user_id 멤버변수 추가

- findCartByUserId를 쿼리 메서드로 사용하기 위해 Cart 엔티티에 읽기 전용 user_id 필드 추가
- `@Column` 어노테이션에 insertable, updatable 값을 false로 지정하여 읽기 전용 필드로 지정

no-cy commented:
피드백 사항들 반영 완료하여 머지 진행하도록 하겠습니다. 💪
